### ISSUE TITLE
MAINT: checking match of current amallela_test_ml_workflow_2 branch workflow against results in DTRA paper

### ISSUE THREAD

**Issue by tylerjereddy** [lanl/ldrd_virus_work#39]
One important thing I need to do is verify that running the DTRA workflow on the current `amallela_test_ml_workflow_2` branch roughly reproduces/matches the results/relevant figures currently in the draft DTRA manuscript.

I recently ran this incantation on `gp160`: `time python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column IN --mapping-method jurgen_schmidt --n_seeds 20`

and it produced the FIC plot below the fold.

<details>

<img width="2400" height="3000" alt="Image" src="https://github.com/user-attachments/assets/e498e9dd-16d0-477a-9235-9b7a55b3c1ba" />

</details>

I was initially panicked because it didn't match the integrin target FIC plot currently in the draft manuscript (Figure 1c). However, it seems Fig 1c was produced with only surface exposed proteins considered in the design matrix, so let me try re-running the workflow on `gp160` again with that adjustment (incantation below). If that ends up matching reasonably well with figure 1c, then I think it would make sense to re-run several more scenarios/target workflows on `gp160`/latest `amallela_test_ml_workflow_2` branch so that I'm well-positioned to double check or update numbers/fill in gaps in the manuscript to get it finalized.

`time python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column IN --mapping-method jurgen_schmidt --n_seeds 20 -s surface_exposed`

Started that run at ~5:58 pm this evening in `gp160` screen session.

**Comment by tylerjereddy** [comment]
> Started that run at ~5:58 pm this evening in gp160 screen session.

Surprisingly, the surface exposed workflow incantation above is still running the next day (September 25th) at 9:40 am. Looks like `lightgbm` is still doing its thing with the most recent output in the screen session: `[LightGBM] [Warning] No further splits with positive gain, best gain: -inf`.

I'm not sure why it seems so much slower with surface exposed filtering active, now more than twice as slow as the original incantation above (the first incantation above completed on the same machine, `gp160`, in about 7.3 hours).

For now, I'll allow it to continue to run--it is important to make sure that I can reproduce the FIC plot in i.e., Figure 1c first before committing too heavily to the results/direction of the manuscript text/re-running of other workflow incantations.

**Comment by tylerjereddy** [comment]
The surface exposed only (integrin target) workflow finally completed in `1283m20.257s` (21.4 hours). I'll try to perform a visual inspection of FIC plot vs. Figure 1c later this evening.

**Comment by tylerjereddy** [comment]
I checked -- the "surface exposed only" integrin target workflow incantation produces an FIC plot with considerable overlap to Figure 1c (see below the fold).  Some kmer rankings got shuffled a bit, but things are close enough that I'm not too worried.

<details>

<img width="2400" height="3000" alt="Image" src="https://github.com/user-attachments/assets/c42730a5-2448-496b-99b9-9768385b0332" />

</details>

The ROC plots that get produced also seem "ok."

So I've started the sialic acid and IgSF "surface exposed only" workflow runs on `gp160`:

- `time python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column SA --mapping-method jurgen_schmidt --n_seeds 20 -s surface_exposed`
- `time python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column IG --mapping-method jurgen_schmidt --n_seeds 20 -s surface_exposed`

**Comment by tylerjereddy** [comment]
At the moment, both of those runs/incantations are still going on `gp160` (looks like this may take a while still...):

<img width="2378" height="93" alt="Image" src="https://github.com/user-attachments/assets/5ab2337f-4b55-4399-973f-a7cc99f4a85a" />

<img width="2378" height="93" alt="Image" src="https://github.com/user-attachments/assets/e07c10e3-22b3-4a25-9261-1b9278117d0a" />

**Comment by tylerjereddy** [comment]
Still running as of this evening as well, this is ridiculously slow.

**Comment by tylerjereddy** [comment]
Code is still running in a `gp160` `screen` session--this is unacceptably slow IMO. It is currently doing work with LightGBM for both target columns. I bet there are some design issues with the looping across estimators/seeds or concurrency settings.

**Comment by tylerjereddy** [comment]
The `SA` and `IG` incantations started at https://github.com/lanl/ldrd_virus_work/issues/39#issuecomment-3336563670 a few days ago have finally completed in the `gp160` `screen` session, in `5702m48.501s` (95 hours) and `5898m54.690s` (98 hours), respectively.

I don't have the bandwidth to check the output results just yet, but there is almost certainly a code performance issue somewhere--perhaps some extra work is happening that doesn't need to because of merge conflict resolution complications, etc.

**Comment by tylerjereddy** [comment]
I performed the SA and IG analysis output visual inspections for the latest runs on branch `amallela_test_ml_workflow_2` on `gp160`. The new SA FIC plot matches reasonably well to Figure S4b in the current version of the draft manuscript, modulo a little shuffling in the exact rank order. The other supporting plots for SA also seem "ok."

Likewise for the new IG FIC plot and Fig S4d, though the shuffling seems more pronounced in this case. The supporting plots for IG are a bit more dubious--several ROC folds end up near random chance.

That's mostly where I thought things would be. The workflow performance/runtime is atrocious though--there must be some unnecessary work happening.

That said, my bandwidth is fairly limited here. The draft manuscript lacks cohesion, clarity, a crisp story, etc., so condensing it down/fixing it to something I could submit *somewhere* is likely to be my main focus.

**Comment by adamwitmer** [comment]
I ran the workflow on the branch `amallela_test_ml_workflow_2` locally on my `ARM M2 MacBook` with the following incantation for the`sialic acid` target column:

```
time python ../viral_seq/run_workflow.py --cache extract --cache-tarball dtra_cache.tar.gz --optimize none --train-file receptor_training.csv --test-file none --workflow DTRA --target-column SA --mapping-method jurgen_schmidt --n_seeds 20 -s surface_exposed
```

and it completed in a reasonable time, i.e.:

```
python ../viral_seq/run_workflow.py --cache extract --cache-tarball   none     8797.01s user 4261.73s system 147% cpu 2:27:16.73 total
```

I am running it again on `Mac` for the `IgSF` target column, as well as on `gp160` for the `Sialic acid` target to double check if I can reproduce the poor workflow performance there.

**Comment by tylerjereddy** [comment]
> on my ARM M2 MacBook

I don't think my slow performance reports were on ARM mac above. You'll probably want to use the same configuration I had, Python `3.13.2` on `gp160` with `pip freeze` showing:

<details>

```
alembic==1.16.4
attrs==25.3.0
biom-format==2.1.16
biopython==1.85
black==23.3.0
certifi==2025.8.3
charset-normalizer==3.4.3
click==8.1.7
cloudpickle==3.1.1
colorlog==6.9.0
contourpy==1.1.0
coverage==7.10.3
cycler==0.11.0
decorator==5.2.1
ete3==3.1.3
execnet==2.1.1
filelock==3.18.0
fonttools==4.40.0
fsspec==2024.10.0
greenlet==3.2.4
h5py==3.14.0
hypothesis==6.137.3
idna==3.10
iniconfig==2.1.0
joblib==1.5.1
jsonschema==4.25.0
jsonschema-specifications==2025.4.1
kiwisolver==1.4.6
legacy-cgi==2.6.3
lightgbm==4.5.0
llvmlite==0.44.0
Mako==1.3.10
markdown-it-py==3.0.0
MarkupSafe==3.0.2
matplotlib==3.10.5
mdurl==0.1.2
msgpack==1.1.1
mypy==1.3.0
mypy_extensions==1.1.0
natsort==8.4.0
numba==0.61.2
numpy==2.1.0
nvidia-nccl-cu12==2.27.7
optuna==4.0.0
packaging==25.0
pandas==2.3.1
pandas-stubs==2.0.2.230605
pathspec==0.12.1
patsy==1.0.1
pillow==11.0.0
platformdirs==4.3.8
pluggy==1.6.0
polars==1.11.0
protobuf==6.31.1
pyarrow==18.0.0
Pygments==2.19.2
pyparsing==3.1.0
pytest==8.4.1
pytest-cov==6.2.1
pytest-mock==3.14.1
pytest-xdist==3.8.0
python-dateutil==2.9.0.post0
pytz==2025.2
PyYAML==6.0.2
ray==2.48.0
referencing==0.36.2
requests==2.32.4
rich==13.4.2
rich-click==1.8.3
rpds-py==0.27.0
ruff==0.8.1
scikit-bio==0.6.2
scikit-learn==1.6.1
scipy==1.16.1
seaborn==0.13.2
shap==0.47.1
shapely==2.0.6
six==1.17.0
slicer==0.0.8
sortedcontainers==2.4.0
SQLAlchemy==2.0.43
statsmodels==0.14.5
taxonomy_ranks==0.0.10
tensorboardX==2.6.2.2
threadpoolctl==3.6.0
tqdm==4.67.1
types-pytz==2023.3.0.0
types-seaborn==0.13.2.20250111
types-shapely==2.0.0.20241221
types-tqdm==4.65.0.1
typing_extensions==4.14.1
tzdata==2025.2
urllib3==2.5.0
-e git+ssh://git@lisdi-git.lanl.gov:10022/treddy/ldrd_virus_work.git@9d6e8a66c0304ebecd050b8376a3c6dc85241e37#egg=viral_seq
xgboost==3.0.3

```

</details>

hash: 9d6e8a66

**Comment by tylerjereddy** [comment]
Adam confirmed in an email that the performance was indeed slow on Linux/`gp160`--in general, information should be shared here for team continuity. I haven't heard anything about a potential solution to this over the past few months--will likely need to investigate myself.

**Comment by tylerjereddy** [comment]
@adamwitmer what's the ETA on addressing this? I have a pretty long list of items to address for this paper and if the turnaround is two months per item I'll need to find another way to get this across the line.

**Comment by adamwitmer** [comment]
I will take a look at the performance issue this weekend.

**Comment by adamwitmer** [comment]
I was able to get the workflow to run in a reasonable time by manually setting the number of threads per core for `xgboost` and `lgbm` classifiers with the diff:

```diff
--- a/viral_seq/run_workflow.py
+++ b/viral_seq/run_workflow.py
@@ -519,7 +519,7 @@ def train_clfr(
     for clfr_name, subdict in classifier_parameters.items():
         clfr = subdict["clfr"]
         clfr.set_params(**subdict["params"], random_state=random_state)
-
+        
         clfr_preds: dict[int, Any] = {}
         feature_count[f"Clfr_{clfr_name}"] = 0
         feature_count[f"SHAP_{clfr_name}"] = 0
@@ -2175,7 +2175,7 @@ if __name__ == "__main__":
             },
             "XGBClassifier": {
                 "clfr": XGBClassifier(),
-                "params": {"n_estimators": 10000, "n_jobs": -1},
+                "params": {"n_estimators": 10000, "n_jobs": -1, "nthread": 16},
             },
             # LGBM parameters determined using suggestion from:
             # https://stackoverflow.com/questions/71285022/why-lightgbm-python-package-gives-bad-prediction-using-for-regression-task
@@ -2187,6 +2187,8 @@ if __name__ == "__main__":
                     "boosting_type": "dart",
                     "n_estimators": 1000,
                     "n_jobs": -1,
+                    "num_threads": 16,
+                    "verbosity": -1,
                 },
             },
         }
```

which produced the following output on `gp160`:


```
RandomForestClassifier Folds: 100%|████████████████████████████████████| 5/5 [1:12:43<00:00, 872.79s/it]
ExtraTreesClassifier Folds: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [1:28:37<00:00, 1063.45s/it]
XGBClassifier Folds: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [32:44<00:00, 392.82s/it]
LGBMClassifier Folds: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [01:57<00:00, 23.57s/it]
```

**Comment by tylerjereddy** [comment]
Do you think `nthread` is the correct way to set the concurrency in XGBoost vs. using `n_jobs` when using the sklearn style API? What was your reasoning for setting both `n_jobs` and `nthread` at the same time instead of changing `n_jobs`? Do you think that makes it clear for me/the reader, or could it maybe be confusing? Can you point me to the reference material you used for making the decision to use `nthread` adjustment instead of `n_jobs` adjustment, and in particular for setting them both at the same time?

After clarifying that, then suggest a way to do this automatically/portably, instead of manual hacking.

**Comment by tylerjereddy** [comment]
@adamwitmer ping on query above -- this paper is going to take quite some time if you can't respond to my queries about what you did and why in a timely manner.

**Comment by tylerjereddy** [comment]
gh-41 was opened in an attempt to resolve the matter of slow performance, given the lack of responsiveness and need to move things forward on this project.

**Comment by tylerjereddy** [comment]
The primary purpose of this issue was to verify that I can match the FIC plots in the draft "DTRA" paper using the latest version of the `amallela_test_ml_workflow_2` feature branch. I established that I can, modulo a bit of random shuffling, so I'll probably close this.

I'll open a separate issue about the performance matter, though I doubt I have the bandwidth or team members ready to root cause that--it seems likely related to load balance/concurrency, as things slow down when there is less work to do (surface exposed filtering has happened) relative to the number of procs it seems.
